working/333 - shared memory multithreading/proposal.md

```diff
@@ -0,0 +1,561 @@
+* `TypedData` - which makes all typed data types shareable.
+  * **BREAKING CHANGE**: making `TypedData` shareable changes the behaviour of `SendPort` which will stop copying it. It's unclear if we want to maintain old behaviour for compatibility reasons. One option here is to say that `TypedData` is only shared when sending through `SendPort` iff it is a member of another explicitly `Shareable` type.    
+* `Pointer` 
+
+## Shared Isolates
+
+Lets take another look at the following example: 
+
+```dart
+int global = 0; 
+
+void main() async {
+  global = 42;
+  await Isolate.run(() {
+    print(global);  // => 0
+    global = 24;
+  });  
+  print(global);  // => 42
+}
+```
+
+Stripped to the bare minimum the example does not seem to behave in a confusing way: it seems obvious that each isolate has its own version of `global` variable and mutations of `global` are not visible across isolates. However, in the real world code such behavior might be hidden deep inside a third party dependency and thus much harder to detect and understand. This behavior also makes interoperability with native code more awkward than it ought to be: calling Dart requires an isolate, something that native code does not really know or care about. 
+
+```dart
+int global;
+
+@pragma('vm:entry-point')
+void foo() => global++;
+```
+
+`@shared` global variables allow developers to tackle this problem - but hidden dependency on global state might introduce hard to diagnose and debug bugs. 
+
+We propose to tackle this problem by introducing the concept of _shared isolate_: **code running in a _shared isolate_ can only access `@shared` state and not any of isolated state, an attempt to access isolated state results in a dynamic `IsolationError`**.
+
+``` dart
+int global = 0; 
+
+@shared
+int sharedGlobal = 0;
+
+void main() async {
+  global = 42;
+  sharedGlobal = 42;
+  await Isolate.run(() {
+    print(global);  // IsolationError: Can't access 'global' when running in shared isolate
+    global = 24;  // IsolationError: Can't access 'global' when running in shared isolate
+
+    print(sharedGlobal);  // => 42
+    sharedGlobal = 24;
+  }, shared: true);  
+  print(global);  // => 42
+  print(sharedGlobal);  // => 24
+}
+```
+
+### Why not compile time isolation?
+
+It is tempting to try introducing a compile time separation between functions which only access `@shared` state and functions which can access isolated state. 
+
+One obvious approach is to introduce a modifier (e.g. `@shared`) which can be applied to function declarations and impose a number of restrictions that `@shared` functions have to satisfy. These restrictions should guarantee that `@shared` functions can only access `@shared` state.
+
+```dart 
+@shared
+void foo() {
+  // ...
+}
+```
+
+* You can only pass subtypes of `Shareable` to `@shared` methods and you can only get a `Shareable` result back.
```

Comment:
It is not strictly speaking necessary - but you can't really do anything useful with value which are not shareable - because you can't touch them from `shared` code. Non-sharable types are opaque in shared code.